### ISSUE TITLE
FR: unified max_tokens and frequency_penalty controls for autoregressive / LLM-style decoders

### ISSUE THREAD

**Issue by grider-transwithai** [CrispStrobe/CrispASR#120]
## Summary
I’d like to request generic generation controls for CrispASR backends that use an autoregressive / LLM-style decoder:
- generated-token cap: `max_tokens` / `max_new_tokens`
- opt-in `frequency_penalty`
The immediate motivation is Cohere ASR, but the request is broader: any backend that decodes by repeatedly selecting the next token from logits could benefit from the same control surface.
## Use Case
In a downstream transcription/subtitle workflow, we use a generated-token cap as a deterministic signal that a request may have entered repeated-loop degradation.
Typical workflow:
1. Run transcription with a generated-token cap.
2. If decoding reaches the cap before EOS, classify the result as likely degenerate/repeated-loop output.
3. Retry the same chunk with a modest `frequency_penalty`.
4. Keep defaults unchanged for normal ASR, where repetition may be valid speech.
This is useful for ASR/AST backends with LLM-style decoders because repeated-loop degradation can otherwise look like a long-running or low-quality transcription rather than a clean failure mode.
## Current Behavior
CrispASR already has related pieces:
- CLI has `-n` / `--max-new-tokens`.
- Some LLM-style backends appear to honor `params.max_new_tokens`.
- The OpenAI-compatible chat endpoint has `max_tokens`.
- Several backends already have temperature/sampling plumbing.
- Some TTS paths already have repetition-penalty precedents.
However, this does not appear to be exposed consistently across all autoregressive / LLM-style decoder backends, and the OpenAI-compatible `/v1/audio/transcriptions` endpoint does not currently accept `max_tokens` or `frequency_penalty`.
### 1. Unified generated-token cap
For backends with autoregressive decoders, please allow users to set a generated-token cap.
Suggested API surface:
- CLI: `-n` / `--max-new-tokens`
- Server `/v1/audio/transcriptions`: `max_tokens`
- Optional server alias: `max_new_tokens`
- C/session API: a shared setter or decode params field
Suggested behavior:
- Stop when EOS is reached or generated-token cap is reached.
- Preserve current defaults when unset.
- Expose stop reason where possible:
  - `stop_reason: "eos"`
  - `stop_reason: "max_tokens"`
  - `truncated: true`
Even a CLI warning when the cap is reached would be useful.
### 2. Unified frequency penalty
For logits-based autoregressive decoders, please add an opt-in `frequency_penalty`.
Suggested API surface:
- CLI: `--frequency-penalty F`
- Server `/v1/audio/transcriptions`: `frequency_penalty`
- C/session API: shared setter or decode params field
Suggested semantics:
- Default `0.0`, disabled.
- Positive values penalize already-generated token IDs proportional to prior count.
- Apply before argmax/sampling.
- Prefer applying only to generated output tokens, not prompt/control tokens.
- If a backend cannot support this cleanly, warn or ignore based on existing capability conventions.
A simple backend-level implementation could be:
```cpp
adjusted_logits[token_id] -= frequency_penalty * generated_count[token_id];
```
### Scope
This does not need to cover every backend initially.
A practical rollout could be:
1. Add shared params / CLI / server fields.
2. Wire support into generic autoregressive decode helpers where available.
3. Add backend-specific support for decoders with custom loops, starting with Cohere.
4. Leave CTC/TDT/non-autoregressive backends unchanged or capability-gated.

### Why This Should Be Opt-In
ASR can contain real repetition:
- stutters
- repeated words
- lyrics/chants
- repeated names
- conversational fillers

So frequency_penalty should not be enabled by default. The goal is not to remove all repetition, but to give clients a recovery knob after max_tokens indicates likely decoder degradation.

**Comment by CrispStrobe** [comment]
sure can do. thanks for the suggestions. update: should mostly work now. could you test it?

**Comment by CrispStrobe** [comment]
Already shipped. `--max-new-tokens N` (or `-n N`) and `--frequency-penalty F` are available in the CLI and wired through all autoregressive backends. Session API: `crispasr_session_set_max_new_tokens()` / `crispasr_session_set_frequency_penalty()`. Server: `max_tokens` field in the JSON request body. See docs/cli.md 'LLM-backend specific' section.

**Comment by grider-transwithai** [comment]
Thanks!
